### ISSUE TITLE
Logloss

### ISSUE THREAD

**Issue by solidate** [rojagtap/abstractive_summarizer#2]
How to get the logloss score for the prediction?

**Comment by rojagtap** [comment]
Try this:
```
@tf.function
def test_step(inp, tar):
    tar_inp = tar[:, :-1]
    tar_real = tar[:, 1:]

    enc_padding_mask, combined_mask, dec_padding_mask = create_masks(inp, tar_inp)

    predictions, _ = transformer(
         inp, tar_inp, 
         True, 
         enc_padding_mask, 
         combined_mask, 
         dec_padding_mask
    )
    loss = loss_function(tar_real, predictions)
    test_loss(loss)

```

I am assuming you have the sequences in the same format as mine, or you can adjust the `tar_inp` and `tar_real` part.
The `test_loss` is the same as the `train_loss` from my code i.e. a `tf.keras.metrics.Mean` instance.

Moreover, you can call this function in the training loop as a validation step to monitor validation loss while training.

**Comment by solidate** [comment]
Sorry. I meant, how to get the softmax values as mentioned in the original paper?

**Comment by rojagtap** [comment]
Umm, still not sure about what exactly you meant, but if you want the softmax output from the logits, use `tf.nn.softmax(logits)`.
Perhaps you can highlight the part from the paper that you are trying to point out?

**Comment by solidate** [comment]
yeah i want the logits only. from which block can i get the logits for the prediction?

**Comment by rojagtap** [comment]
```
predictions, _ = transformer(
     inp, tar_inp, 
     True, 
     enc_padding_mask, 
     combined_mask, 
     dec_padding_mask
)

```

This returns the logits for the generated sequence of shape (batch_size, sequence_length, vocab_size)

**Comment by solidate** [comment]
predictions[: ,-1:, :]
In this what are the three dimensions? 
Sorry for asking so many doubts

**Comment by rojagtap** [comment]
No issues! I'll try to keep this as simple as possible:

While training, we put the complete target sequence as the decoder input, so for each word to be predicted, at the decoder, we have the original word as the decoder input instead of the supposedly incorrectly predicted word (I am assuming you know seq2seq basics). It's called [Teacher Forcing](https://towardsdatascience.com/what-is-teacher-forcing-3da6217fed1c)

However, in the inference (prediction or testing) step, we do not have the expected output to be fed in the decoder, so we have to use the previously predicted word as the decoder input for the next word prediction.

So in the code, as and when we predict tokens (words), we append them to the decoder inputs. So the variable `predictions` contains the complete predicted sequence so far. But we only need the last word here; the `-1` indicates the last token of the predicted sequence. To elaborate even further, consider a question answering task with the same model instead of summarization:

Q: What is your name?
Step 1:
Enc Input: What is your name?
Dec Input: <go>
Output: <go> My                     ... (Hence we only take the last word)
Step 2:
Enc Input: What is your name?
Dec Input: <go> My
Output: <go> My name
Step 3:
Enc Input: What is your name?
Dec Input: <go> My name
Output: <go> My name is
...

Hope you got it :)
